### PR TITLE
CORDA-2576 Viral update propagation: select attachments from those loaded by node on startup.

### PR DESCRIPTION
**Pull request by josecoll**
Adjust the way we select attachments so we don't use the latest known version but rather, the installed version (stop viral propagation)

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 1]
```diff
@@ -467,7 +436,7 @@ open class TransactionBuilder(
         require(isReference || constraints.none { it is HashAttachmentConstraint })
 
         val minimumRequiredContractClassVersion = stateRefs?.map { services.loadContractAttachment(it).contractVersion }?.max() ?: DEFAULT_CORDAPP_VERSION
```
Comment (szymonsztuka) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: `minimumRequiredContractClassVersion` is obsolete now

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 2]
```diff
@@ -467,7 +436,7 @@ open class TransactionBuilder(
...
-        return services.attachments.getLatestContractAttachments(contractClassName, minimumRequiredContractClassVersion).firstOrNull()
+        return services.cordappProvider.getContractAttachmentID(contractClassName)
                 ?: throw MissingContractAttachments(states, contractClassName, minimumRequiredContractClassVersion)
```
Comment (szymonsztuka) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: ... and this may become `throw MissingContractAttachments(states, contractClassName)`

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt [hunk 3]
```diff
@@ -234,10 +234,14 @@ data class SignedTransaction(val txBits: SerializedBytes<CoreTransaction>,
             // TODO - should this be a [TransactionVerificationException]?
             val missingClass = requireNotNull(e.message) { "Transaction $ltx is incorrectly formed." }
 
-            val attachment = requireNotNull(services.attachments.internalFindTrustedAttachmentForClass(missingClass)) {
+            val attachmentId = requireNotNull(services.cordappProvider.getContractAttachmentID(missingClass)) {
                 "Transaction $ltx is incorrectly formed. Could not find local dependency for class: $missingClass."
             }
 
+            val attachment = requireNotNull(services.attachments.openAttachment(attachmentId)) {
```
Comment (szymonsztuka) [inline core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: Should it get only trusted attachments (loaded by 'app' or 'rpc' only), as `internalFindTrustedAttachmentForClass` used to do?

Comment (josecoll) [reply core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: The previous `requireNotNull` check will fail if the CorDapp is not loaded from the local filesystem (eg. it must be trusted if this line of code executes)

Comment (szymonsztuka) [reply core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: indeed!

File: node/src/integration-test/kotlin/net/corda/contracts/SignatureConstraintVersioningTests.kt [hunk 4]
```diff
@@ -83,57 +79,6 @@ class SignatureConstraintVersioningTests {
         assertNotNull(stateAndRef)
         assertEquals(transformetMessage, stateAndRef!!.state.data.message)
     }
-
-    @Test
-    fun `cannot evolve from higher contract class version to lower one`() {
```
Comment (szymonsztuka) [inline node/src/integration-test/kotlin/net/corda/contracts/SignatureConstraintVersioningTests.kt]: What about preserving this test but make it positive - `can evolve ...`+ not expecting the exception?

**Comment by mikehearn** [comment]
Is this right? I was expecting the change in question to be in TransactionBuilder, where we select an attachment to satisfy constraints or pick constraints for new issuances - but it seems this one changes the backwards compatibility path that's used if a class is referenced but not attached?

**Comment by josecoll** [comment]
I have made the change in the TransactionBuilder 
https://github.com/corda/corda/pull/4745/files#diff-16ba09b8cf423b43ec89c8ed3220848aR433 for the constraints checking.
Are you saying this is the only place it should be effected?

**Comment by mikehearn** [comment]
Ah sorry, I overlooked that by mistake. I'll do another review and try to read it more carefully now.

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt [hunk 5]
```diff
@@ -234,10 +234,14 @@ data class SignedTransaction(val txBits: SerializedBytes<CoreTransaction>,
             // TODO - should this be a [TransactionVerificationException]?
             val missingClass = requireNotNull(e.message) { "Transaction $ltx is incorrectly formed." }
 
-            val attachment = requireNotNull(services.attachments.internalFindTrustedAttachmentForClass(missingClass)) {
+            val attachmentId = requireNotNull(services.cordappProvider.getContractAttachmentID(missingClass)) {
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: This code here is not related to the propagation of jar updates.
This is a hack to address transactions that were created in version 3 without all the required dependencies.
Also it was supposed to work not only for contracts.

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 6]
```diff
@@ -175,15 +175,15 @@ open class TransactionBuilder(
             val missingClass = e.message
             requireNotNull(missingClass) { "Transaction is incorrectly formed." }
 
-            val attachment = services.attachments.internalFindTrustedAttachmentForClass(missingClass!!)
+            val attachmentId = services.cordappProvider.getContractAttachmentID(missingClass!!)
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Same as the change in SignedTransaction.

File: node/src/integration-test/kotlin/net/corda/contracts/SignatureConstraintVersioningTests.kt (hunk 4 above)
Comment (josecoll) [reply node/src/integration-test/kotlin/net/corda/contracts/SignatureConstraintVersioningTests.kt]: I tried that but you end up with a related exception (that will also disappear when you back out the non-downgrade rule check):

`net.corda.core.contracts.TransactionVerificationException$TransactionVerificationVersionException: No-Downgrade Rule has been breached for contract class net.corda.contracts.DummyMessageContract. The output state contract version '2' is lower than the version of the input state '3'., transaction: 2B8342431B9F07E7C28687A39F9ACEE7948C16CA872A35FE96E7CC933BB7B8B3`

File: core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt (hunk 5 above)
Comment (josecoll) [reply core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: Not for this PR but ...
Isn't this a security breach then?
ie. I could upload any attachment with malicious code to my Node and get the trusted Contract JAR to classload it by reference.

Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: Why would you want hack your own node this way?

Comment (josecoll) [reply core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: to bypass verification perhaps.

Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: It is your node. 
If you want to bypass verification on an invalid transaction that nobody else would accept you can do that in many ways.
But there is nothing to gain in doing that.

Comment (mikehearn) [reply core/src/main/kotlin/net/corda/core/transactions/SignedTransaction.kt]: Right. As a reminder the threat model currently does *not* include malicious *installed* apps. Only malicious contracts and peers.

**Comment by josecoll** [comment]
Rebased from master (no conflicts).
